### ISSUE TITLE
The argument type 'ModalRoute<Object?>?' can't be assigned to the parameter type 'PageRoute<dynamic>'

### ISSUE THREAD

**Issue by tjx666** [flutter/flutter#80749]
Copy following code to latest flutter version: 2.0.5 project:

```dart
import 'package:flutter/material.dart';

final routeObserver = RouteObserver<PageRoute>();

class HomePage extends StatefulWidget {
  HomePage({Key? key}) : super(key: key);

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> with RouteAware {
  AppLifecycleState? _notification;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    routeObserver.subscribe(this, ModalRoute.of(context));
  }

  @override
  void dispose() {
    routeObserver.unsubscribe(this);
    super.dispose();
  }

  @override
  void didPush() {
    // Route was pushed onto navigator and is now topmost route.
    print('[HomePage] route did push!');
  }

  @override
  void didPopNext() {
    // Covering route was popped off the navigator.
    print('[HomePage] route did pop next!');
  }

  void goToSecondPage() {
    Navigator.of(context).push(MaterialPageRoute(
      builder: (context) {
        return SecondPage();
      },
    ));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('home page'),
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Text('app lifeCycle state: $_notification'),
            ElevatedButton(
              onPressed: goToSecondPage,
              child: Text('go to second page'),
            )
          ],
        ),
      ),
    );
  }
}
```

![image](https://user-images.githubusercontent.com/41773861/115293159-90fca500-a189-11eb-8a09-9c61cf14b21f.png)

**Comment by stuartmorgan** [comment]
The error is correct; you can't directly assign a superclass to variable whose type is a subclass, since that's not typesafe, nor can you assign a nullable value to a non-nullable variable, for the same reason.

**Comment by tjx666** [comment]
@stuartmorgan  but the code is copy from [official document](https://api.flutter.dev/flutter/widgets/RouteObserver-class.html#widgets.RouteObserver.1), and I can search useful help info about how to fix it.

**Comment by xu-baolin** [comment]
Fixed via #77606
